metrics/web/test/dashboard/domain/entities/metrics/receive_build_day_project_metrics_updates_test.dart

```diff
@@ -0,0 +1,308 @@
+// Use of this source code is governed by the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:metrics/common/domain/entities/build_day.dart';
+import 'package:metrics/common/domain/repositories/build_day_repository.dart';
+import 'package:metrics/dashboard/domain/entities/collections/date_time_set.dart';
+import 'package:metrics/dashboard/domain/entities/metrics/build_day_project_metrics.dart';
+import 'package:metrics/dashboard/domain/entities/metrics/build_performance.dart';
+import 'package:metrics/dashboard/domain/usecases/parameters/project_id_param.dart';
+import 'package:metrics/dashboard/domain/usecases/receive_build_day_project_metrics_updates.dart';
+import 'package:mockito/mockito.dart';
+
+import '../../../../test_utils/matchers.dart';
+
+void main() {
+  group("ReceiveBuildDayProjectMetricsUpdates", () {
+    const projectId = 'id';
+    const projectIdParam = ProjectIdParam(projectId);
+    const collectionEquality = DeepCollectionEquality();
+
+    final buildDayRepository = _BuildDayRepositoryMock();
+    final useCase = ReceiveBuildDayProjectMetricsUpdates(buildDayRepository);
+
+    PostExpectation<Stream<List<BuildDay>>> whenStreamBuildDays() {
+      return when(buildDayRepository.projectBuildDaysInDateRangeStream(
+        projectId,
+        from: anyNamed('from'),
+      ));
+    }
+
+    BuildDay createBuildDay({
+      int successful = 0,
+      int failed = 0,
+      int unknown = 0,
+      int inProgress = 0,
+      Duration totalDuration = Duration.zero,
+      DateTime day,
+    }) {
+      return BuildDay(
+        projectId: projectId,
+        successful: successful,
+        failed: failed,
+        unknown: unknown,
+        inProgress: inProgress,
+        totalDuration: totalDuration,
+        day: day ?? DateTime.now(),
+      );
+    }
+
+    Matcher buildDayMetricsMatcher({
+      Matcher projectIdMatcher = anything,
+      Matcher buildNumberMetricMatcher = anything,
+      Matcher performanceMetricMatcher = anything,
+    }) {
+      return allOf(
+        projectIdMatcher,
+        buildNumberMetricMatcher,
+        performanceMetricMatcher,
+      );
```

Comment:
Great idea of using the `allOf` matcher 👍 
But what if we'll try to simplify using this `buildDayMetricsMatcher`. Currently, we have to pass a `predicate` that explicitly accesses the concrete value of the `BuildDayProjectMetrics`. How about using the `isA` matcher with the `having` construction as follows:
 ```suggestion
      final metricsMatcher = isA<BuildDayProjectMetrics>();
      return allOf([
        metricsMatcher.having(
          (metrics) => metrics.projectId,
          'projectId',
          projectIdMatcher,
        ),
        metricsMatcher.having(
          (metrics) => metrics.buildNumberMetric,
          'buildNumberMetric',
          buildNumberMetricMatcher,
        ),
        metricsMatcher.having(
          (metrics) => metrics.performanceMetric,
          'performanceMetric',
          performanceMetricMatcher,
        ),
      ]);
```
This allows passing matches for concrete fields and their values instead of using a predicate every time. For example, in the appropriate test cases, instead of: 
```dart
        final projectIdMatcher = predicate<BuildDayProjectMetrics>(
          (metrics) => metrics.projectId == expectedProjectId,
        );

        expect(
          metricsStream,
          emits(buildDayMetricsMatcher(projectIdMatcher: projectIdMatcher)),
        );
```
we can write: 
```dart
        final metricsMatcher = buildDayMetricsMatcher(
          projectIdMatcher: equals(expectedProjectId),
        );

        expect(metricsStream, emits(metricsMatcher));
```

What do you think?